src/python.cpp

```diff
@@ -2207,3 +2207,91 @@ SEXP py_eval_impl(const std::string& code, bool convert = true) {
   return result;
 
 }
+
+// [[Rcpp::export]]
+SEXP py_convert_pandas_df(PyObjectRef obj) {
+  
+  // pd.DataFrame.items() iterates over (column name, Series) pairs
+  PyObjectPtr items(PyObject_CallMethod(obj, "items", NULL));
+
+  List list;
+  
+  while (true) {
+
+    // get next tuple
+    PyObjectPtr tuple(PyIter_Next(items));
+    if (tuple.is_null()) {
+      if (PyErr_Occurred())
+        stop(py_fetch_error());
+      else
+        break;
+    }
+    
+    // access Series in slot 1
+    PyObjectPtr series(PySequence_GetItem(tuple, 1));
+    
+    // and extract dtype
+    PyObjectPtr dtype(PyObject_GetAttrString(series, "dtype"));
+    PyObjectPtr dtype_str(PyObject_GetAttrString(dtype, "name"));
+    
+    // as well as values
+    PyObjectPtr values(PyObject_GetAttrString(series, "values"));
+    
+    SEXP sexp;
```

Comment:
nit: for safety, maybe initialize this with `R_NilValue`? (It looks like all paths below will set this but it's nice to "know" that it will never have an invalid or undefined state)

Reply:
Or consider an Rcpp object (e.g. RObject) so the underlying R object would automatically be protected for you.

Reply:
Using Rcpp in all cases now, this will be more readable than lots of protection/unprotection I guess...